presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,532 @@
+    {
+        OGCGeometry geometry = deserialize(input);
+        String type = geometry.geometryType();
+        if (type.equals(POLYGON)) {
+            SpatialReference reference = geometry.getEsriSpatialReference();
+            Envelope envelope = new Envelope();
+            geometry.getEsriGeometry().queryEnvelope(envelope);
+            Point centroid = new Point((envelope.getXMin() + envelope.getXMax()) / 2, (envelope.getYMin() + envelope.getYMax()) / 2);
+            return serialize(OGCGeometry.createFromEsriGeometry(centroid, reference));
+        }
+        throw new IllegalArgumentException("st_centroid only applies to polygon. Input type is: " + type);
+    }
+
+    @Description("Returns count of coordinate components")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stCoordinateDimension(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).coordinateDimension();
+    }
+
+    @Description("Returns spatial dimension of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.BIGINT)
+    public static long stDimension(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).dimension();
+    }
+
+    @SqlNullable
+    @Description("Returns true if and only if the line is closed")
+    @ScalarFunction("st_is_closed")
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stIsClosed(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        if (geometry.geometryType().equals(LINE) || geometry.geometryType().equals(MULTILINE)) {
+            MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+            int pathCount = lines.getPathCount();
+            boolean result = true;
+            for (int i = 0; result && i < pathCount; i++) {
+                Point start = lines.getPoint(lines.getPathStart(i));
+                Point end = lines.getPoint(lines.getPathEnd(i) - 1);
+                result = result && end.equals(start);
+            }
+            return result;
+        }
+        throw new IllegalArgumentException("st_is_closed only support line and multiline. Input type is: " + geometry.geometryType());
+    }
+
+    @SqlNullable
+    @Description("Returns true if and only if the geometry is empty")
+    @ScalarFunction
+    @SqlType(StandardTypes.BOOLEAN)
+    public static Boolean stIsEmpty(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).isEmpty();
+    }
+
+    @Description("Returns the length of line")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stLength(@SqlType(GEOMETRY) Slice input)
+    {
+        return deserialize(input).getEsriGeometry().calculateLength2D();
+    }
+
+    @Description("Returns the maximum X coordinate of geometry")
+    @ScalarFunction
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stMaxX(@SqlType(GEOMETRY) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
```

Comment:
nit: the logic of getting an envelope for a geometry is repeated in multiple functions; would it make sense to extract it to a helper method?